### PR TITLE
fix: resolve false positive constructor arguments

### LINKED ISSUE #1873: There was an error compiling your contract.
`0x68d0657245d7bb542ce8910a49e34d1ad12e792b` fails to verify. There is nothing special about this contract to warrant it not to verify.

* 0x57d1585168bccafd2475a53cc8b519b78cdf5fa0
* 0x68d0657245d7bb542ce8910a49e34d1ad12e792b
* 0x3b005f89571d44b9335b0b397997165e605fa6e6
* 0x22ecd603fdb52d59090ffb9e6fb3b67ad0a29d4f

### Steps to reproduce

1. Attempt to verify `0x68d0657245d7bb542ce8910a49e34d1ad12e792b`
2. Receive the message `There was an error compiling your contract.`

### Expected behaviour

Verify Contract

### Actual behaviour

Error

### PR DESCRIPTION
**Pull request by zachdaniel**
Resolves #1873 

### Bug Fixes
* We only want to take everything up to the swarm hash.

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [x] I checked whether I should update the docs and did so if necessary

### CODE REVIEW

**Comment by goodsoft** [comment]
Are you _really_ sure there can't be any false positives _after_ the `0029` we need? 
If the constructor arguments are there, are you sure they can't contain `0029` themselves?
Or, even, `x0 02 9x`, which would match as well?

**Comment by acravenho** [comment]
@zachdaniel @goodsoft Here is an example of a false positive where we would run into an issue: https://etherscan.io/address/0x64d14595152b430cf6940da15c6e39545c7c5b7e#code

Please see the constructor arguments and the `0029`.

**Comment by zachdaniel** [comment]
@goodsoft definitely not really sure. I only just started looking into it this morning, and had a similar thought myself, but I figured I would get the bare bones solution for that specific contract up in front of people.

**Comment by zachdaniel** [comment]
Okay, so the swarm source starts with a known value, I think. So perhaps what we do is skip to the swarm source and split there?

**Comment by zachdaniel** [comment]
https://solidity.readthedocs.io/en/v0.4.21/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode

**Comment by acravenho** [comment]
Yes, that's correct. Are you able to identify the swarm source without the `0029`?

**Comment by acravenho** [comment]
Nice! That should work.

**Comment by zachdaniel** [comment]
We could use the length + `bzzr0` bytes? Although I think its implying that the mapping itself might change at anytime.

**Comment by zachdaniel** [comment]
I think it looks reliable enough for now. If they "support other ways to retrieve contract metadata in the future", this can be updated.

**Comment by acravenho** [comment]
{`a165627a7a72305820`} - {32 bytes} -  {`0029`} - {`constructor arguments`}

**Comment by acravenho** [comment]
Comparing now against these addresses:

* 0xada931ac30aff9e51148d6aaf565fc89e7e90bc0 ✔️ 
* 0x8f639224f992e466ecdffc3f5880194a73f1d93b
* 0x1155c5605d08585594a428f473c77eeb33f28bc1
* 0x71320e08c267746616456aeca4efde73ed98caf3
* 0x99831db28f7bb7c03a1a6c517a54ac3e80da2be1
* 0x7c6ef33ab884862dadf412f7271d669fd3cb8875
* 0x5152345f9722bc1bef5fd4ec0428946d5eb18c82
* 0xe241db15493adfe3c49884255f4edf012389bcc8
* 0x61fe6d3692abdb010347d89d7ec391b719115519
* 0xbf67ebc297685863ae7e0e6f88eaffe8052a0d3b
* 0x74dd3ee13dc6714de5670198f6e5cd586690b79a
* 0x69868c621f229fcacfab4b2238b30607c3a61ec6
* 0x348c19f7de78acf3c89c7a5fd67b3f4b515bf69a
* 0x2020fe35fc3a2ad06073b5083917b284f927f7b3
* 0x9975bdf4826b1280f413084441f95ca59da40aea
* 0x7f4f073b00de185ff737252e2614f11cba980e83
* 0x7c7c4a72a0995389cf4324b5727bf6dd442afc10

**Comment by acravenho** [comment]
The two contracts that failed above don't seem to have a swarm source hash. Looks like they were created prior to when swarm source was created.

**Comment by zachdaniel** [comment]
Do they also have constructor arguments? If so, then we'll need another heuristic for those. I'm almost done with the code to use the swarm hash.

**Comment by acravenho** [comment]
No constructor arguments.  They fail when trying to verify them but probably for some other reason. Looks like an edge case.

**Comment by zachdaniel** [comment]
Well for now, if we don't find a swarm hash I'll have it use the whole thing.

**Comment by zachdaniel** [comment]
Fighting with the data that we are currently testing against. I don't think it conforms to the standard and as such led me on a wild goose chase.

File: apps/explorer/test/explorer/smart_contract/verifier_test.exs [hunk 1]
```diff
@@ -181,13 +181,29 @@ defmodule Explorer.SmartContract.VerifierTest do
       swarm_source = "3c381c1b48b38d050c54d7ef296ecd411040e19420dfec94772b9c49ae106a0b"
 
       bytecode =
-        "0x608060405234801561001057600080fd5b5060df8061001f6000396000f3006080604052600436106049576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff16806360fe47b114604e5780636d4ce63c146078575b600080fd5b348015605957600080fd5b5060766004803603810190808035906020019092919050505060a0565b005b348015608357600080fd5b50608a60aa565b6040518082815260200191505060405180910390f35b8060008190555050565b600080549050905600a165627a7a72305820"
```
Comment (zachdaniel) [inline apps/explorer/test/explorer/smart_contract/verifier_test.exs]: It was asserting that part of the swarm metadata hash would come back, but that would be incorrect.

**Comment by zachdaniel** [comment]
Alright @vbaranov @acravenho @goodsoft this should be ready to go for real now.

File: apps/explorer/lib/explorer/smart_contract/verifier.ex [hunk 2]
```diff
@@ -86,17 +86,28 @@ defmodule Explorer.SmartContract.Verifier do
...
+    do_extract_bytecode(<<>>, String.downcase(code))
+  end
 
-    bytecode
+  defp do_extract_bytecode(extracted, remaining) do
+    case remaining do
+      <<>> ->
+        String.reverse(extracted)
+
+      "a165627a7a72305820" <> <<_::binary-size(64)>> <> "0029" <> _constructor_arguments ->
```
Comment (ayrat555) [inline apps/explorer/lib/explorer/smart_contract/verifier.ex]: what is `a165627a7a72305820`?

File: apps/explorer/lib/explorer/smart_contract/verifier.ex [hunk 3]
```diff
@@ -86,17 +86,28 @@ defmodule Explorer.SmartContract.Verifier do
...
-      |> List.first()
-      |> String.split_at(-64)
+    do_extract_bytecode(<<>>, String.downcase(code))
+  end
 
-    bytecode
+  defp do_extract_bytecode(extracted, remaining) do
+    case remaining do
+      <<>> ->
+        String.reverse(extracted)
```
Comment (ayrat555) [inline apps/explorer/lib/explorer/smart_contract/verifier.ex]: `String.reverse` reverses graphemes but it looks like you extract bytes from the string. I think one grapheme can contain multiple bytes. Maybe we can use a list as an accumulator to resolve this issue?

File: apps/explorer/lib/explorer/smart_contract/verifier.ex (hunk 2 above)
Comment (zachdaniel) [reply apps/explorer/lib/explorer/smart_contract/verifier.ex]: It’s what the swarm hash starts with.

File: apps/explorer/lib/explorer/smart_contract/verifier.ex (hunk 3 above)
Comment (zachdaniel) [reply apps/explorer/lib/explorer/smart_contract/verifier.ex]: Yeah, that is a better idea. I almost went with the route of deciding the input and operating on the raw bytes, maybe I should just go that route. Either way, I agree a list accumulator and then reversed and cast to a binary would be best.

File: apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex [hunk 4]
```diff
@@ -7,20 +7,21 @@ defmodule Explorer.SmartContract.Verifier.ConstructorArguments do
...
-      creation_input_data
-      |> String.split(data_with_swarm)
-      |> List.last()
-      |> String.replace("0x", "")
+  defp extract_constructor_arguments("a165627a7a72305820" <> <<_::binary-size(64)>> <> "0029" <> constructor_arguments) do
+    constructor_arguments
+  end
 
-    expected_arguments_data == arguments_data
+  defp extract_constructor_arguments(<<_::binary-size(1)>> <> rest) do
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: I would use size 2 here to align with bytes _just in case_.

Comment (zachdaniel) [reply apps/explorer/lib/explorer/smart_contract/verifier/constructor_arguments.ex]: Ah, yeah great point :+1:

**Pull request merged**